packages/eslint-plugin-react-hooks/src/RulesOfHooks.js

```diff
@@ -100,6 +100,16 @@ function isInsideComponentOrHook(node) {
   return false;
 }
 
+function isInsideDoWhileLoop(node) {
+  while (node) {
+    if ( node.type === 'DoWhileStatement' ) {
+      return true;
+    }
+    node = node.parent;
+  }
+  return false;
+}
+
```

Comment:
Adding an extra utility to traverse up the tree to check if the node is called within a `do/while` loop.

Reply:
yesterday was the first time i looked at the code in this plugin. i'm curious why all the "is hook inside loop" logic isn't handled this way. seems like the same could be done for all types of loops/cycles and invalid code.  i found the "counting segments" code somewhat hard to follow and this function is much more clear to me. maybe the counting segments was done for performance reasons?

without fully understanding the existing code, it seems like this plugin could be rewritten so that we first find all "hook" functions, then call a "isInsideCycle" that climbs up the tree looking for nodes like `DoWhileStatement` or `ForStatement` etc.

anyways, good work on this (and thanks)!

Reply:
I agree - it does feel quite complex at first glance, and my initial expectation was that these would be handled top-down, with loops and control structures at the top, and then function/hook calls at the bottom, but I assumed there was a good reason (likely performance) for it to be that way. Anyway, rewriting that would be a major overhaul of the plugin, and one definitely outside of this PR's scope.